### DISCUSSION TITLE
Running brew-installed utilities with sudo

### DISCUSSION THREAD

**Discussion by gmile** [Homebrew/discussions#3575] [Linux] [answered]
What's a reasonable way to use `sudo` together with utilities, installed via `brew`?

I've installed and successfully using command-line utilities via `brew` on Kubuntu 22.04.1 Sometimes, I need to edit files that require elevated permissions. For example, I need to run `sudo nvim /etc/sudoers`, and such.

Unfortunately, an attempt to do this renders an error:

```
$ sudo nvim /etc/sudoers
$ sudo: nvim: command not found
$
```

I've identified at least 3 different ways to work around this error:

1. modify what's known as [`secure_path` variable](https://superuser.com/a/927599/39098). doing so results in modification of root's `PATH` variable, so `sudo` starts to "see" binaries installed by homebrew,

2. (haven't tried this) wipe out `secure_path` variable so that current user's `PATH` is used to locate binary, 

3. provide a full path to binary every time I use it, e.g.:

    ```
    sudo /home/linuxbrew/.linuxbrew/bin/nvim /etc/sudoers
    ```

Is there a better, "officially" acknowledged or documented, more reasonable way of running commands installed using brew - using `sudo`?

P.S. I've looked through existing discussions using [this filter](https://github.com/Homebrew/discussions/discussions/categories/linux?discussions_q=category%3ALinux+sudo) and [this filter](https://github.com/Homebrew/discussions/discussions/categories/linux?discussions_q=category%3ALinux+root). [This reply](https://github.com/Homebrew/discussions/discussions/2716#discussioncomment-1904467) was insightful. though haven't looked through issues or PRs for homebrew.

**Comment by SMillerDev** [comment]
None of this is brew related though, you'll be better off asking the sudo developers for recommendations.

**Comment by danielnachun** [comment]
Oddly I would expect `sudo -E` to work here, but it seems like it doesn't.  However `sudo env PATH=$PATH nvim` does seem to work, so you might be able to just alias `sudo` as `sudo env PATH=$PATH`.  There may be security implications to always passing your `PATH` through to `sudo` like this so use with care.

**Reply by gromgit** [reply]
> Oddly I would expect `sudo -E` to work here, but it seems like it doesn't.

Yeah, that only works if the `setenv` flag is set in `/etc/sudoers`. That's why the `sudo` man page says:

> `-E`, `--preserve-env`
>   Indicates to the security policy that the user wishes to preserve their existing environment variables.  The security policy may return an error if the user does not have permission to preserve the environment.

`sudo` can say **NO ENV FOR YOU!** even if you ask nicely. :grinning:

**Comment by carlocab** [comment, marked as answer]
If all you need is `nvim`, it may suffice to add `/home/linuxbrew/.linuxbrew/opt/neovim/bin` to your `secure_path`, so that you can do `sudo nvim` without allowing `sudo` to be executed with everything in `/home/linuxbrew/.linuxbrew/bin`.
